### PR TITLE
[KGA-87] fix: high offset memory expansion computations

### PR DESCRIPTION
**Pull request by enitrat**
Fixes computation on high memory offsets.
The remediation proposed by the audit finding was not sufficient enough. This one looks fine, although a bit more costly (even though uint256_add is not that computationally intensive)

- - -
This change is [Reviewable](https://reviewable.io/reviews/kkrt-labs/kakarot/1604)

### CODE REVIEW

File: tests/utils/constants.py [hunk 1]
```diff
@@ -48,6 +48,8 @@
 BLOCK_NUMBER = 0x42
 BLOCK_TIMESTAMP = int(time())
 
+FELT_252_PRIME = 0x800000000000011000000000000000000000000000000000000000000000001
+
```
Comment (ClementWalter) [inline tests/utils/constants.py:52]: use `from starkware.cairo.lang.cairo_constants import DEFAULT_PRIME`

**Review by ClementWalter — CHANGES_REQUESTED** [review]
use `from starkware.cairo.lang.cairo_constants import DEFAULT_PRIME`

File: cairo_zero/tests/src/kakarot/test_gas.py [hunk 2]
```diff
@@ -33,32 +36,37 @@ def test_should_return_correct_expansion_cost(
...
-            size_2=integers(min_value=0, max_value=0xFFFFF),
+            offset_1=integers(min_value=0, max_value=DEFAULT_PRIME - 1),
+            size_1=integers(min_value=0, max_value=DEFAULT_PRIME - 1),
+            offset_2=integers(min_value=0, max_value=DEFAULT_PRIME - 1),
+            size_2=integers(min_value=0, max_value=DEFAULT_PRIME - 1),
         )
         def test_should_return_max_expansion_cost(
             self, cairo_run, offset_1, size_1, offset_2, size_2
         ):
+            MEMORY_COST_U32 = 0x200018000000
```
Comment (ClementWalter) [inline cairo_zero/tests/src/kakarot/test_gas.py:47]: could use instead `calculate_memory_gas_cost(2**32 - 1)`

File: cairo_zero/tests/src/kakarot/test_gas.py [hunk 3]
```diff
@@ -33,32 +36,37 @@ def test_should_return_correct_expansion_cost(
...
-                    [
-                        (offset_1, size_1),
-                        (offset_2, size_2),
-                    ],
-                ).cost
+
+            # If the memory expansion is greater than 2**27 words of 32 bytes
+            # We saturate it to the hardcoded value corresponding the the gas cost of a 2**32 memory size
+            expected_saturated = (
+                MEMORY_COST_U32 if expansion.expand_by > 2**32 else expansion.cost
```
[the remark this answers is not in the document]
Comment (enitrat) [reply cairo_zero/tests/src/kakarot/test_gas.py:67]:
```
        let (q, _) = unsigned_div_rem(max_offset, 2 ** 32);
        tempvar range_check_ptr = range_check_ptr;
        jmp expansion_cost_saturated if q != 0;
```

q = 2**32 => q = 1 => saturated donc oui

**Review by ClementWalter — APPROVED** [review]

**Pull request merged**